src/storage/dictionary_encoding_tile.cpp

```diff
@@ -0,0 +1,186 @@
+
+#include <cstdio>
+#include <sstream>
+
+#include "catalog/schema.h"
+#include "common/exception.h"
+#include "common/macros.h"
+#include "type/serializer.h"
+#include "common/internal_types.h"
+#include "type/ephemeral_pool.h"
+#include "concurrency/transaction_manager_factory.h"
+#include "storage/backend_manager.h"
+#include "storage/tile.h"
+#include "storage/tile_group_header.h"
+#include "storage/tuple.h"
+#include "storage/tuple_iterator.h"
+#include "storage/dictionary_encoding_tile.h"
+
+namespace peloton {
+namespace storage {
+
+DictEncodedTile::DictEncodedTile(BackendType backend_type, TileGroupHeader *tile_header,
+     const catalog::Schema &tuple_schema, TileGroup *tile_group,
+     int tuple_count) : Tile(backend_type, tile_header, tuple_schema, tile_group, tuple_count),
+     is_dict_encoded(false),
+     original_schema(tuple_schema) {
+	std::vector<catalog::Column> columns;
+	size_t offset = 0;
+
+	// find out which column to encode
+	for (oid_t i = 0; i < column_count; i++) {
+		auto column_type = schema.GetType(i);
+		original_schema_offsets.emplace(offset, i);
+		if (column_type == type::TypeId::VARCHAR ||
+				column_type == type::TypeId::VARBINARY) {
+			catalog::Column encoded_column(type::TypeId::TINYINT,
+				type::Type::GetTypeSize(type::TypeId::TINYINT),
+				schema.GetColumn(i).GetName(), true);
+			columns.push_back(encoded_column);
+			dict_encoded_columns.insert(i);
+		} else {
+			columns.push_back(schema.GetColumn(i));
+		}
+		offset += schema.GetLength(i);
+	}
+
+	// if nothing needs to be encoded
+	if (dict_encoded_columns.empty()) return;
+
+	delete[] data;
+	schema = catalog::Schema(columns);
+	tuple_length = schema.GetLength();
+	tile_size = num_tuple_slots * tuple_length;
+	data = new char[tile_size];
+}
+
+// delete data is enough
+DictEncodedTile::~DictEncodedTile(){}
+
+type::Value DictEncodedTile::GetValue(const oid_t tuple_offset, const oid_t column_id) {
+	if (dict_encoded_columns.count(column_id) > 0) {
+		// encoded
+		auto idx_val = Tile::GetValue(tuple_offset, column_id);
```

Comment:
Consider renaming this variable. It wasn't immediately clear what idx_val was. Consider changing "idx_val" to "encoded_val" and "idx" to "dict_index".